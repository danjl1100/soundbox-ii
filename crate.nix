{
  pkgs,
  system,
  craneLib,
  advisory-db,
  cargoExtraArgs ? "",
  extraBuildArgs ? {},
  pname ? null,
  src ? null,
  srcDir ? ./.,
} @ inputs: let
  src =
    if (builtins.isNull inputs.src)
    then (craneLib.cleanCargoSource srcDir)
    else inputs.src;

  # Common arguments can be set here to avoid repeating them later
  commonArgs =
    {
      inherit src cargoExtraArgs;

      buildInputs =
        [
          # Add additional build inputs here
        ]
        ++ pkgs.lib.optionals pkgs.stdenv.isDarwin [
          # Additional darwin specific inputs can be set here
          pkgs.libiconv
          pkgs.darwin.apple_sdk.frameworks.CoreServices
        ];
    }
    // (
      if (builtins.isNull pname)
      then {}
      else {inherit pname;}
    );

  # Build *just* the cargo dependencies, so we can reuse
  # all of that work (e.g. via cachix) when running in CI
  cargoArtifacts = craneLib.buildDepsOnly commonArgs;

  # Build the actual crate itself, reusing the dependency
  # artifacts from above.
  my-crate = craneLib.buildPackage (commonArgs
    // {
      inherit cargoArtifacts;
    }
    // extraBuildArgs);

  my-crate-doc = craneLib.cargoDoc (commonArgs
    // {
      inherit cargoArtifacts;
      cargoDocExtraArgs = "--workspace --no-deps"; # override default which is "--no-deps"
    });
in rec {
  checks = {
    # Build the crate as part of `nix flake check` for convenience
    inherit my-crate;

    inherit my-crate-doc;

    # Run clippy (and deny all warnings) on the crate source,
    # again, resuing the dependency artifacts from above.
    #
    # Note that this is done as a separate derivation so that
    # we can block the CI if there are issues here, but not
    # prevent downstream consumers from building our crate by itself.
    my-crate-clippy = craneLib.cargoClippy (commonArgs
      // {
        inherit cargoArtifacts;
        cargoClippyExtraArgs = "--all-targets";
        # TODO: deny warnings (?)
        # cargoClippyExtraArgs = "--all-targets -- --deny warnings";
      });

    # Check formatting
    my-crate-fmt = craneLib.cargoFmt {
      inherit src;
    };

    # TODO re-enable audits to ensure they pass!
    # # Audit dependencies
    # my-crate-audit = craneLib.cargoAudit {
    #   inherit src advisory-db;
    # };

    # Run tests with cargo-nextest
    # Consider setting `doCheck = false` on `my-crate` if you do not want
    # the tests to run twice
    my-crate-nextest = craneLib.cargoNextest (commonArgs
      // {
        inherit cargoArtifacts;
        partitions = 1;
        partitionType = "count";
        # TODO: enable code coverage, only if it's worth it
        # } // pkgs.lib.optionalAttrs (system == "x86_64-linux") {
        #   # NB: cargo-tarpaulin only supports x86_64 systems
        #   # Check code coverage (note: this will not upload coverage anywhere)
        #   my-crate-coverage = craneLib.cargoTarpaulin (commonArgs // {
        #     inherit cargoArtifacts;
        #   });
      });
  };

  package = my-crate;
  doc = my-crate-doc;

  drv-open-doc-for-crate = crate-name:
    pkgs.writeShellScriptBin "open-doc-${crate-name}" ''
      ${pkgs.xdg-utils}/bin/xdg-open "file://${my-crate-doc}/target/doc/${crate-name}/index.html"
    '';

  devShellFn = inputs:
    craneLib.devShell (inputs
      // {
        inherit checks;
      });
}
