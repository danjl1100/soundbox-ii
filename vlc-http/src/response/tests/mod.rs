// Copyright (C) 2021-2024  Daniel Lambert. Licensed under GPL-3.0-or-later, see /COPYING file for details

#![allow(clippy::panic)] // that's what tests are supposed to do!

use crate::Response;
use std::{
    io::BufReader,
    path::{Path, PathBuf},
};
use test_log::test;
use tracing::{error, info};

#[test]
fn parse() -> Result<(), Box<dyn std::error::Error>> {
    let input_folder = {
        let mut path_buf: PathBuf = std::env::current_exe()?
            .ancestors()
            .nth(4)
            .expect("test exe within target/debug/deps/vlc_http-????")
            .to_path_buf();
        // crate root
        path_buf.extend(&["vlc-http", "src", "response", "tests", "input"]);
        path_buf
    };

    for entry in std::fs::read_dir(input_folder)? {
        let entry = entry?;
        if entry.metadata()?.is_file() {
            let name = entry.file_name();
            let name = name.to_string_lossy();
            let path = entry.path();

            info!(%name, path=%path.display(), "test input");

            parse_file(&name, &path)?;
        } else {
            error!(path=%entry.path().display(), "invalid file type");
            panic!("invalid file type for {}", entry.path().display());
        }
    }
    Ok(())
}

fn parse_file(name: &str, path: &Path) -> Result<(), Box<dyn std::error::Error>> {
    let file = std::fs::File::open(path)?;
    let reader = BufReader::new(file);

    let response = Response::from_reader(reader)?;
    insta::assert_ron_snapshot!(name, response);

    Ok(())
}
