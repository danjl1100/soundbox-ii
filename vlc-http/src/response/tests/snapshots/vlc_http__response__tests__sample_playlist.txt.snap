---
source: vlc-http/src/response/tests/mod.rs
expression: response
---
Response(
  inner: PlaylistInfo(Info(
    items: [
      Item(
        duration_secs: None,
        id: "1580",
        name: "Floaters.mp3",
        url: "file:///path/to/Music/Jimmy%20Fontanez/Floaters.mp3",
      ),
    ],
  )),
)
